### ISSUE TITLE
ON CLUSTER doesn't work with hosts in multiple shards

### ISSUE THREAD

**Issue by vavrusa** [ClickHouse/ClickHouse#1287]
The queries across the whole cluster don't work when a host has multiple shards:

```
DROP TABLE test ON CLUSTER 'cluster1'

click1	9000	371	An error occured before execution: Code: 371, e.displayText() = DB::Exception: There are two exactly the same ClickHouse instances click1:9000 in cluster cluster1, e.what() = DB::Exception	32	0
```

The host has this indeed three replicas of different shards:

```
SELECT *
  FROM system.clusters
 WHERE host_name = 'click1'


┌─cluster──┬─shard_num─┬─shard_weight─┬─replica_num─┬─host_name─┬─host_address─┬─port─┬─is_local─┬─user────┬─default_database─┐
│ cluster1 │        28 │            1 │           1 │ click1    │ 10.0.10.100  │ 9000 │        0 │ default │ r0               │
│ cluster1 │        29 │            1 │           2 │ click1    │ 10.0.10.100  │ 9000 │        0 │ default │ r1               │
│ cluster1 │        30 │            1 │           3 │ click1    │ 10.0.10.100  │ 9000 │        0 │ default │ r2               │
└──────────┴───────────┴──────────────┴─────────────┴───────────┴──────────────┴──────┴──────────┴─────────┴──────────────────┘
```

There's really no reason for this check to exist, if you want to make sure it's not a configuration error, you should perhaps use host_name:port/default_database for uniqueness check?

cc @ludv1x

**Comment by ludv1x** [comment]
Hi!
Indeed, for the simplicity of the first distributed DDL version, I intentionally skipped implementation of the case when a server has several shards in different databases. I thought it is not the widespread case.

It is not complicated to add support of this case.

**Comment by vavrusa** [comment]
Ah I see, supporting this use case would be wonderful 👌

**Comment by vavrusa** [comment]
So my understanding is that only ALTER queries for replicated tables are problematic, since some require you to execute the query only on leader replica, and some require you to execute the query on any single replica.

For CREATE | ATTACH | DROP | DETACH it should be executed on distinct server instances regardless of how many shards do they have. For example in my case, I have 3 shards in databases r{0,1,2}, but if write:

```
DROP TABLE r0.test ON CLUSTER 'cluster1'
```

I expect r0.test to be dropped on all hosts, and I don't expect other replicas or shards to be affected.

**Comment by ajardan** [comment]
We also have multiple shards and host a replica of each shard on every node, which basically makes ON CLUSTER requests unusable...

**Comment by filimonov** [comment]
https://github.com/yandex/ClickHouse/blob/6aa2313decf619faf62c04c7bc8a7d6913831140/dbms/src/Interpreters/DDLWorker.cpp#L73

**Comment by ghuname** [comment]
Can we expect this feature in nearby future?

**Comment by touseefrafique** [comment]
We are also facing this issue. Does anyone knows any plans on the fix

**Comment by unflag** [comment]
Same issue. We have to truncate some tables periodically and it would be much more convenient if our apps wouldn't know anything about cluster shards and their tables - only cluster name.

**Comment by etah000** [comment]
I have finished a patch to support distributed ddl(currently, only delete/updatet/drop partition operations have been tested) on cross replication clusters。 [distributed-alter-query](https://github.com/etah000/ClickHouse/tree/distributed-alter-query)。

**Comment by filimonov** [comment]
Fixed in  #11508 & #11703
